python/ray/data/_internal/planner/exchange/push_based_shuffle_task_scheduler.py

```diff
@@ -561,20 +631,35 @@ def _map_partition(
         *map_args: List[Any],
     ) -> List[Union[BlockMetadata, Block]]:
         mapper_outputs = map_fn(idx, block, output_num_blocks, *map_args)
-        meta = mapper_outputs.pop(-1)
 
-        parts = []
+        # A merge task may produce results for multiple downstream reducer
+        # tasks. Therefore, each map task should give each merge task a
+        # partition of its outputs, where the length of the partition is equal
+        # to the number of reducers downstream to the merge task.
+        partition = []
         merge_idx = 0
-        while mapper_outputs:
-            partition_size = schedule.get_num_reducers_per_merge_idx(merge_idx)
-            parts.append(mapper_outputs[:partition_size])
-            mapper_outputs = mapper_outputs[partition_size:]
-            merge_idx += 1
-        assert len(parts) == schedule.num_merge_tasks_per_round, (
-            len(parts),
+        num_reducers_for_cur_merger = schedule.get_num_reducers_per_merge_idx(merge_idx)
+        for output in mapper_outputs:
+            if len(partition) == num_reducers_for_cur_merger:
+                yield partition
+
+                partition = []
+                merge_idx += 1
+                num_reducers_for_cur_merger = schedule.get_num_reducers_per_merge_idx(
+                    merge_idx
+                )
+
+            if num_reducers_for_cur_merger is None:
+                assert not partition
+                # The last output should be the metadata.
+                yield output
```

Comment:
also assert the type of output is metadata, and it is the last one in mapper_outputs?